### ISSUE TITLE
Detect log-spam in tests

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#109297]
When adding log statements in Kubernetes it can sometimes be difficult to tell just how frequent a code path is. For example: https://github.com/kubernetes/kubernetes/pull/109181, or any of about 500 issues & PRs: https://github.com/kubernetes/kubernetes/issues?q=log+spam

I'd like to add infrastructure to all our test frameworks (unit, integration & E2E) to detect log spam. Things that would be interesting to look at:
- Total log volume - works best for unit tests; e2e & integration with variable run times are likely to have too much variance on total volume
- Logging rates (percentiles) - less dependent on test length, probably needs a way to separate out logging during cluster bring-up
- Repeated log lines - typically indicates a low-information or high-spam log

The reasons we should care about log spam:
1. It obscures informative log information
2. It can fill up disks and/or saturate log pipelines

/sig testing instrumentation
/area logging

**Comment by tallclair** [comment]
Thinking about this a bit more, I think this is unlikely to be particularly useful for unit tests, and we should focus on the integration & E2E cases instead.

**Comment by pohly** [comment]
For integration testing this could be built into the test. All code that we care about runs as part of the test. The advantage is that we then can also collects stats for different verbosity levels even when the test itself only runs at one level.

For E2E testing we want to check logs of components in the cluster, not of the E2E test suite, because the code running in the suite is limited (primarily client-go). That means we would need to piggy-back onto the log collection at the end of an E2E run and do some log post-processing.

Besides checking for log spam, that could also be used to check for error messages. An error message may or may not be normal, so it's a bit questionable how to surface that there were errors.

**Comment by tallclair** [comment]
If we add log metrics (https://github.com/kubernetes/klog/issues/167), this would be much easier to do. We could even diff the metrics before & after each test to identify if a particular test triggered log spam (only for serial tests).

**Comment by pohly** [comment]
I've started experimenting with log collection after an E2E test run in https://github.com/kubernetes/kubernetes/pull/133844.

The log is available and can be parsed. Where I am struggling is with defining a sane policy for what characterizes log spam such that it can be detected automatically and reliably. Perhaps "log entries/s within a certain time interval (e.g. five minutes) > threshold"? The failure message then can include the log entries in that interval.

The threshold must depend on the log verbosity. Should it also be different by component or do we expect the same sane behavior from all of them? I'm leaning towards the same threshold for all of them.

**Comment by pohly** [comment]
> identify if a particular test triggered log spam (only for serial tests)

It's not practical to run all the E2E tests sequentially, it takes too long. But this poses a problem for the "log entries/s" rule because log volume is going to be proportional to the current load.

**Comment by pohly** [comment]
> The threshold must depend on the log verbosity. 

If the log entries included the verbosity level (they do for JSON, but not for text output), then a single E2E run could check different thresholds for different levels. But the problem with -v4 (the default) was that the post-run check only gets the tail of the full log because too much log data was produced, so currently I am using -v2.

Up to which [verbosity level](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use) do we care about log spam? I'd say up to  up to 3 ("extended information about changes") but not 4 ("debug level").

**Comment by tallclair** [comment]
The log output is also going to depend on which tests were run, not just how many. I think the ideal way to approach this would be to get some baseline for the log volume for each test job, and flag test runs that deviate from that baseline by more than a certain amount (e.g. 5%).

> Up to which verbosity level do we care about log spam? I'd say up to up to 3 ("extended information about changes") but not 4 ("debug level").

Yeah, I agree we don't care about logspam at V4. I could take it or leave it at V3, but I'm probably biased since GKE usually runs at V2. I don't think klog actually records verbosity level though? I think it's just error/warning/info?

**Comment by BenTheElder** [comment]
> Yeah, I agree we don't care about logspam at V4. I could take it or leave it at V3, but I'm probably biased since GKE usually runs at V2. I don't think klog actually records verbosity level though? I think it's just error/warning/info?

v4 is what we run most of our CI at IIRC, it is for `kind` jobs at least.

**Comment by BenTheElder** [comment]
I think metrics would be more reliable and cheaper than attempting to fetch the entire logs and scan them in the e2e tests.

We already typically attempt to persist logs to $ARTIFACTS, so we could also instead check there ad-hoc.

... I'm going to continue writing up https://github.com/kubernetes/enhancements/issues/5468

**Comment by pohly** [comment]
>  I think the ideal way to approach this would be to get some baseline for the log volume for each test job, and flag test runs that deviate from that baseline by more than a certain amount (e.g. 5%).

Are you trying to reduce log output *in the tests* (typically integration tests) or *in the components under test* (kube-apiserver, kube-scheduler, etc.)? While it may be nice to know which test(s) were running, the components should not spam the logs no matter what clients are doing.

> v4 is what we run most of our CI at IIRC, it is for kind jobs at least.

It'll depend on whether we really can check log volume for all jobs or whether we have to limit it to dedicated jobs with a known client behavior.

> I think metrics would be more reliable and cheaper than attempting to fetch the entire logs and scan them in the e2e tests.

I'm not sure how reliably a metric could spot undesirable behavior like spamming the log for a few minutes in a one hour run. Perhaps a histogram for "log volume/second" with different buckets could work - but that then doesn't tell a developer where or when the problem happened. Some analysis of the logs would still be needed, which might not be possible due to log rotation.

XREF: https://github.com/kubernetes/kubernetes/pull/133844#issuecomment-3257172519

**Comment by pohly** [comment]
Let me add that while metrics might work for log spam, they won't work for detecting DATA RACE reports. Those can only be found by inspecting the output.

**Comment by serathius** [comment]
My 2 cents, I don't think the goal should be to try to directly find which log is spamming, instead we want to make sure that log volume doesn't increase drastically between releases. We could measure that on different verbosity levels and tract changes over time.

**Comment by pohly** [comment]
How do we measure log volume so that it is comparable between releases?

**Comment by BenTheElder** [comment]
> Are you trying to reduce log output in the tests (typically integration tests) or in the components under test (kube-apiserver, kube-scheduler, etc.)?

I assume the latter.

There's a reason we don't directly obtain the cluster logs in the e2e framework currently.

Obtaining logs is not a fully portable operation. Kubeadm clusters happen to use static pods and register those kubelets but many clusters don't have registered control plane pods. Setting that aside, pod logs will completely miss kubelet which seems important to include.

Also, fetching all logs from tests has considerable scalability concerns, so we probably at least cannot default enable this check.

Metrics have the benefit of being more portable and constrained in resource usage.

Dumping logs from a 5k node cluster takes considerable resources and we don't want to do it twice.

**Comment by BenTheElder** [comment]
Every cluster management tool / installer / distro has bespoke tooling to get *the cluster logs* because log management and cluster architecture are pretty flexible and pluggable.

User application pods are standardized but again, even if we want to constrain it to architectures that are pod based ... we're giving up kubelet right off the bat.

We could still test the cluster logs, but I think it will require assumptions about the cluster, and I don't think we should fetch the logs twice, so that pushes it into a post-processing step after logs have been dumped.

**Comment by pohly** [comment]
We don't have to have this test working in all cluster configurations to detect regressions. Supporting it for e.g. kind-based jobs would already cover a good portion of our upstream  code.

But that aside, the feasibility of all approaches depends on a better definition of what "log spam" is. Let's first clarify what we want to achieve, i.e. what we want to measure, then we can decide how we actually get that number.

**Comment by BenTheElder** [comment]
> We don't have to have this test working in all cluster configurations to detect regressions. Supporting it for e.g. kind-based jobs would already cover a good portion of our upstream code.

Agreed, but less portable tests are not as suitable for the e2e framework, or at least they need to be opt-in / feature-flagged.

Also, we already fetch the logs, so it might make more sense to process them from the dump we made to ARTIFACTS instead of fetching them twice.

**Comment by pohly** [comment]
Recent incident: https://github.com/kubernetes/kubernetes/issues/135996
